segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentStateStore.java

```diff
@@ -72,40 +89,136 @@
 
     @Override
     public CompletableFuture<SegmentState> get(String segmentName, Duration timeout) {
-        String stateSegment = StreamSegmentNameUtils.getStateSegmentName(segmentName);
+        return findLatestValidState(segmentName, timeout);
+    }
+
+    private CompletableFuture<SegmentState> findLatestValidState(String segmentName, Duration timeout) {
+        String stateSegment1 = StreamSegmentNameUtils.getFirstStateSegmentName(segmentName);
+        String stateSegment2 = StreamSegmentNameUtils.getSecondStateSegmentName(segmentName);
+
+        final CompletableFuture<SegmentState>[] states = new CompletableFuture[2];
+        final SegmentProperties[] props = new SegmentProperties[2];
+
         TimeoutTimer timer = new TimeoutTimer(timeout);
-        return this.storage
-                .getStreamSegmentInfo(stateSegment, timer.getRemaining())
-                .thenComposeAsync(sp -> readSegmentState(sp, timer.getRemaining()), this.executor)
+
+        CompletableFuture<SegmentState> retVal1 = this.storage
```

Comment:
These two blocks are identical. Perhaps a private method is inorder.

Reply:
They are changing two different values out of the scope of lambda. Could not find a cleaner way of doing it. Do you have any recommendations?

Reply:
You can use Pair<SegmentState, SegmentProperties> from apache commons. We already depend on it.